ccm_web/server/src/api/api.admin.handler.ts

```diff
@@ -103,4 +110,104 @@ export class AdminApiHandler {
+      const method = HttpMethod.Post
+      const body = {
+        user: {
+          // API doesn't provide separate given- and surname fields
+          name: `${user.givenName} ${user.surname}`,
+          sortable_name: `${user.surname}, ${user.givenName}`,
+          terms_of_use: true,
+          skip_registration: true
+        },
+        pseudonym: {
+          unique_id: loginId,
+          send_confirmation: false
+        },
+        communication_channel: {
+          type: 'email',
+          address: email,
+          skip_confirmation: true
+        },
+        force_validations: false
+      }
+      logger.debug(`Sending admin request to Canvas endpoint: "${endpoint}"; method: "${method}"; body: "${JSON.stringify(body)}"`)
+      const response = await this.requestor.request<CanvasUser>(endpoint, method, body)
+      logger.debug(`Received response with status code (${String(response.statusCode)})`)
+      const {
+        id,
+        name,
+        sortable_name, // eslint-disable-line
+        short_name, // eslint-disable-line
+        login_id // eslint-disable-line
+      } = response.body
+      return {
+        id,
+        name,
+        sortable_name,
+        short_name,
+        login_id,
+        email
+      }
+    } catch (error) {
+      const errorResponse = handleAPIError(error, `Login ID: ${loginId}; Role: ${user.type}`)
+      return {
+        statusCode: errorResponse.canvasStatusCode,
+        errors: [errorResponse]
+      }
+    }
+  }
+
+  async createExternalUsers (users: SectionExternalUserDto[], accountID: number): Promise<Array<CanvasUser | APIErrorData>> {
+    const NS_PER_SEC = BigInt(1e9)
+    const start = process.hrtime.bigint()
+
+    // Try creating all Canvas users; failure means user already exists
+    const createUserPromises = users.map(async (user) => await this.createExternalUser(user, accountID))
+    const createUserResponses = await Promise.all(createUserPromises)
+
+    const end = process.hrtime.bigint()
+    logger.debug(`Time elapsed to create (${users.length}) external users: (${(end - start) / NS_PER_SEC}) seconds`)
+
+    return createUserResponses
+  }
+
+  async getUserInfo (loginId: string): Promise<CanvasUser | APIErrorData> {
+    const safeLoginId = loginId.replace('@', '+')
+
+    try {
+      const endpoint = `users/sis_login_id:${safeLoginId}`
+      const method = HttpMethod.Get
+      logger.debug(`Sending admin request to Canvas endpoint: "${endpoint}"; method: "${method}"`)
+      const response = await this.requestor.get<CanvasUser>(endpoint)
+      logger.debug(`Received response with status code (${String(response.statusCode)})`)
+      const {
+        id,
+        name,
+        sortable_name, // eslint-disable-line
+        short_name // eslint-disable-line
+      } = response.body
+      return {
+        id,
+        name,
+        sortable_name,
+        short_name,
+        login_id: loginId,
+        email: undefined as any
```

Comment:
Let's not use `any` here; we can create a simpler base interface and have another extend it as `CanvasUserWithEmail`, or we can return the email here too, right?

Reply:
If you or I as admins use this Canvas API function, we will get the user's `login_id` and `email`.  However, the special purpose admin API token I created for this purpose, with limited permissions, cannot access those values.

I didn't feel like creating yet another interface with a `CanvasUser` variant.  Syntax checkers were telling me I needed to return values for those fields, though.  I decided to re-use the `loginId` argument.  I noticed when the value for `email` was `undefined` it would work without appearing in the JSON results.  I had to cast it as `any` to satisfy syntax, though.  Now I see that Codacy's eslint doesn't like that casting.

So…

Here's a new idea.  Let's just return the minimum needed.  What does the UI need?  A Boolean value, like `{"exists":true}`?

Reply:
I've added the new interface `CanvasUserLoginEmail` to use with new API features that need those fields.  I wish this wasn't necessary.  I tried making `email` optional, but that raised other Typescript errors that I didn't know how to resolve.